File: Codigos estudiantes por lenguaje/PY/Bryann Valderrama/Strings/rabinKarp.py
'''Rabin-Karp

   - Busca un patron dentro de un texto.
   - Genera Hashes para substrings en el texto y el patron segun los # de ASCCI de cada caracter.
   - Busca Hashes similares y luego compara los caracteres individuales.
   - Complejidad Tiempo: 
        - Promedio y Mejor Caso: O(n+m)
        - Peor Caso: O(n*m)
'''

from sys import stdin, stdout
wr = stdout.write

D = 256  # Caracteres del ASCII


def RKsearch(pat, txt, q):

    M = len(pat)
    N = len(txt)
    j = 0  # i itera en el texto, j itera en el patron
    h, p, t = 1, 0, 0  # Hashes

    for i in range(M-1):
        h = (h*D) % q
    for i in range(M):
        p = (D * p + ord(pat[i])) % q
        t = (D * t + ord(txt[i])) % q
    for i in range(N-M+1):
        if p == t:
            for j in range(M):
                if txt[i+j] != pat[j]:
                    break
            j += 1
            if j == M:
                wr(f'"{txt[i:i+M]}" encontrado en el indice ({i} - {i+M-1})\n')

        if i < N-M:
            t = (D * (t - ord(txt[i]) * h) + ord(txt[i+M])) % q
            if t < 0:
                t += q


txt = 'ABCABCABC'
pat = 'ABC'
RKsearch(pat, txt, 17)  # Se ingresa num primo

File: Codigos estudiantes por lenguaje/C++/Julian_silva/Strings/rabinKarp.cpp
using namespace std;
#include<bits/stdc++.h>
#include<cstdlib>
int d=256;
void rabinKarp(string pat,string txt,int q){
    int M=pat.size();
    int N=txt.size();
    int i,j;
    int p=0;
    int t=0;
    int h=1;
    for(i=0;i<M-1;i++){
        h=(h*d)%q;
    }
    for(i=0;i<M;i++){
        p=(d*p+pat[i])%q;
        t=(d*i+txt[i])%q;
    }
    for(i=0;i<=N-M;i++){
        if(p==t){
            for(j=0;j<M;j++){
                if(txt[i+j]!=pat[j]){
                    break;
                }
            }
            if(j==M){
                printf("Patron encontrado en el indice %d - %d\n",i,(i+M-1));
            }
        }
        if(i<N-M){
            t=(d*(t-txt[i]*h)+txt[i+M])%q;
            if(t<0){
                t=(t+q);
            }
        }
    }
}
int main() {
    string txt="ANITALAVAATINA";
    string pat="ANITA";
    rabinKarp(pat,txt,d);
}
